File: fastify_server/config/lifecycle/01-app-yaml.mjs
import fp from 'fastify-plugin';
import { AppYamlConfig, AppYamlConfigSDK } from 'app-yaml-static-config';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// This file matches the signature bound by `server.mjs`:
// if (module.onStartup) startupHooks.push(module.onStartup);

export async function onStartup(server, config) {
    console.log("Initializing App Yaml Static Config...");

    // In this context, 'server' is the Fastify instance

    // We can assume config is accessible or derived
    const configDir = process.env.CONFIG_DIR || path.join(__dirname, '..', '..', '..', 'common', 'config');

    server.log.debug({ configDir }, "Initializing Configuration");

    // We register it as a plugin to ensure decorations happen in correct scope if needed,
    // although this hook is called just before listen().

    // Since we are inside an async startup hook, we can just await the initialization
    const appEnv = (process.env.APP_ENV || 'dev').toLowerCase();  // Normalize to lowercase
    await AppYamlConfig.initialize({
        files: [
            path.join(configDir, 'base.yml'),
            path.join(configDir, `server.${appEnv}.yaml`)
        ],
        configDir
    });

    const appConfig = AppYamlConfig.getInstance();
    const sdk = new AppYamlConfigSDK(appConfig);

    server.log.info({ providers: sdk.listProviders() }, 'Configuration loaded');

    // Decorate Fastify instance for global access
    // Note: server.config may already be decorated with bootstrap config by server.mjs
    // We overwrite it with AppYamlConfig instance which has getAll()/toObject() methods
    if (!server.hasDecorator('config')) {
        server.decorate('config', appConfig);
    } else {
        // Overwrite existing bootstrap config with AppYamlConfig
        server.config = appConfig;
    }
    if (!server.hasDecorator('sdk')) {
        server.decorate('sdk', sdk);
    }
}

File: fastapi_server/config/lifecycle/01_app_yaml.py
from fastapi import FastAPI
from app_yaml_static_config import AppYamlConfig, AppYamlConfigSDK
from app_yaml_static_config.types import InitOptions
import os

async def onStartup(app: FastAPI, config: dict):
    print("Initializing App Yaml Static Config...")
    
    # Determine config directory relative to where config files should be
    # For this example, let's assume a 'config' folder at root, which is parent of this lifecycle folder's parent
    # But usually configured via env var or passed in config
    
    config_dir = os.getenv('CONFIG_DIR', os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '..', 'common', 'config'))
    env = os.getenv('APP_ENV', 'dev').lower()  # Normalize to lowercase

    # Ensure config dir exists or fail gracefully/log
    if not os.path.exists(config_dir):
         print(f"Warning: Config dir {config_dir} not found, using defaults")

    options = InitOptions(
        files=[
            os.path.join(config_dir, 'base.yml'),
            os.path.join(config_dir, f'server.{env}.yaml')
        ],
        config_dir=config_dir
    )
    
    AppYamlConfig.initialize(options)
    app_config = AppYamlConfig.get_instance()

    app.state.config = app_config
    app.state.sdk = AppYamlConfigSDK(app_config)

    print(f"App Yaml Config loaded: {app.state.sdk.list_providers()}")
